File: parser/src/test-data/javascript/categories/blocks/control-flow.js
// fixture: cjs/blocks/control-flow.js
// module system: CommonJS  (governing: staging/cjs/package.json, "type": "commonjs")
// nature: runtime-bearing
// syntax floor: ES2018 (for await) for everything except the final class, which
//   needs ES2022 for its static field and static initialisation block. Those
//   two lines are the only above-baseline syntax in the file and can be excised
//   without touching anything else — but CLASS_STATIC_BLOCK is a declared
//   js_block kind AND a declared js_scope kind, so excising them leaves two
//   enum values with no fixture. See MANIFEST.md's syntax-floor table.
//
// Port of java/blocks/ControlFlowExamples.java and NestedBlockLinking.java.
// Every js_block kind that is not an exception form: FUNCTION_BODY, BLOCK, IF,
// ELSE, FOR, FOR_IN, FOR_OF, WHILE, DO, SWITCH, SWITCH_CASE, LABELED,
// CLASS_BODY, CLASS_STATIC_BLOCK, MODULE_BODY.
//
// `label` is a column because TypeScript's enum audit found `outer: for (...)`
// emitting the loop and DROPPING the label — a correctly positioned row with a
// missing sibling, invisible to every count-based check. Labels appear in both
// their forms here: on a loop and on a bare block.
//
// The block/scope distinction the schema draws is visible throughout: a block is
// SYNTAX, a scope is BINDING. A bare `{}` containing only `var` opens no scope,
// and js_block.opensScope says so.

'use strict';

const items = [1, 2, 3];
const obj = { a: 1, b: 2 };

function everyLoopForm(n) {
  const out = [];

  for (let i = 0; i < n; i += 1) { out.push(i); }

  // A for with an empty head in every slot, and one with no body block at all.
  let j = 0;
  for (;;) { if (j++ > 2) { break; } }
  for (let k = 0; k < 2; k += 1) out.push(k);   // no braces: the body is a statement

  // Comma operator in both the init and the update.
  for (let a = 0, b = 10; a < b; a += 1, b -= 1) { out.push(a + b); }

  for (const key in obj) { out.push(key); }
  for (const value of items) { out.push(value); }

  // Destructuring in a for-of head.
  for (const [index, value] of items.entries()) { out.push(index + value); }

  while (out.length < 20) { out.push(0); }

  do { out.pop(); } while (out.length > 15);

  return out;
}

async function asyncLoops(stream) {
  const out = [];
  for await (const chunk of stream) { out.push(chunk); }
  return out;
}

function branching(value) {
  if (value > 10) {
    return 'big';
  } else if (value > 5) {
    return 'medium';
  } else {
    return 'small';
  }
}

// An if with no braces, an else with no braces, and a dangling else — the shape
// §6 of BUILDING-A-PARSER.md warns about in extractor code, present here as
// input rather than as implementation.
function unbraced(value) {
  if (value) return 'yes';
  else return 'no';
}

function nestedDangling(a, b) {
  if (a)
    if (b) return 'both';
    else return 'a only';        // binds to the INNER if
  return 'neither';
}

function switching(kind) {
  switch (kind) {
    case 'a':
    case 'b':
      // Two case labels, one body. Fall-through between them is intentional
      // and invisible.
      return 'ab';
    case 'c': {
      // A braced case body opens a real block AND a scope.
      const local = 'c';
      return local;
    }
    case 'd':
      // Deliberate fall-through with no break.
      kind = 'e';
    case 'e':
      return 'de';
    default:
      return 'other';
  }
}

// A switch with the default in the MIDDLE, which is legal and changes nothing
// about matching order.
function defaultInMiddle(kind) {
  switch (kind) {
    case 1: return 'one';
    default: return 'other';
    case 2: return 'two';
  }
}

// Labels: on a loop, on a nested loop, and on a bare block. `break label` from
// a block is the closest JavaScript has to a goto.
function labelled(matrix) {
  const found = [];
  outer:
  for (const row of matrix) {
    inner:
    for (const cell of row) {
      if (cell === 0) { continue outer; }
      if (cell < 0) { break outer; }
      if (cell === 99) { break inner; }
      found.push(cell);
    }
  }

  block: {
    if (found.length === 0) { break block; }
    found.push(-1);
  }

  return found;
}

// A bare block containing only `var`. It is a BLOCK with opensScope = false —
// the block exists in the syntax and creates no binding scope, which is the
// distinction js_block and js_scope are two relations for.
function bareBlock() {
  {
    var notScoped = 1;
  }
  {
    let scoped = 2;
    return notScoped + scoped;
  }
}

// Four-deep nesting, with a function boundary in the middle so the block tree
// and the scope tree diverge.
function deeplyNested(input) {
  if (input) {
    for (const x of input) {
      while (x > 0) {
        try {
          const inner = () => {
            switch (x) {
              case 1: {
                return 'one';
              }
              default:
                return 'many';
            }
          };
          return inner();
        } finally {
          break;
        }
      }
    }
  }
  return null;
}

class WithStaticBlock {
  static registry = null;          // [ES2022] static class field
  static {                          // [ES2022] static initialisation block
    WithStaticBlock.registry = new Map();
  }
  method() {
    // A class method body is a FUNCTION_BODY block whose parent is a CLASS_BODY.
    { return WithStaticBlock.registry; }
  }
}

module.exports = {
  everyLoopForm, asyncLoops, branching, unbraced, nestedDangling,
  switching, defaultInMiddle, labelled, bareBlock, deeplyNested, WithStaticBlock
};

File: parser/src/test-data/csharp/categories/blocks/ControlFlow.cs
// Port of java/blocks/{ControlFlowExamples,NestedBlockLinking}.java
// Every statement form that opens a block or a scope, four levels deep, plus
// the four C# block forms Java has no spelling for: `lock`, `checked`,
// `unchecked` and the `using` statement / declaration.
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;

namespace Fixtures.Ported.Blocks;

public class ControlFlowExamples
{
    private readonly object gate = new object();

    public int IfElseChain(int value)
    {
        if (value < 0)
        {
            return -1;
        }
        else if (value == 0)
        {
            return 0;
        }
        else
        {
            return 1;
        }
    }

    // Unbraced bodies: a statement that is not a block, in every position that
    // accepts one.
    public int UnbracedBodies(int value)
    {
        if (value < 0)
            return -1;
        else
            value++;

        for (int i = 0; i < 3; i++)
            value += i;

        while (value > 100)
            value -= 10;

        foreach (var c in "abc")
            value += c;

        do
            value--;
        while (value > 50);

        return value;
    }

    public int ForLoops(int[] data)
    {
        int total = 0;

        for (int i = 0; i < data.Length; i++)
        {
            total += data[i];
        }

        // Multiple initialisers, multiple incrementors, no condition body.
        for (int i = 0, j = data.Length - 1; i < j; i++, j--)
        {
            total += data[i] - data[j];
        }

        // Empty header.
        for (; ; )
        {
            break;
        }

        return total;
    }

    public int ForEachForms(IEnumerable<int> source, Dictionary<string, int> map)
    {
        int total = 0;

        foreach (int item in source)
        {
            total += item;
        }

        foreach (var item in source)
        {
            total += item;
        }

        // Deconstructing foreach — no Java form.
        foreach (var (key, value) in map)
        {
            total += key.Length + value;
        }

        foreach (KeyValuePair<string, int> pair in map)
        {
            total += pair.Value;
        }

        return total;
    }

    public int WhileAndDo(int seed)
    {
        int value = seed;

        while (value > 0)
        {
            value /= 2;
        }

        do
        {
            value++;
        }
        while (value < 10);

        return value;
    }

    // Switch STATEMENT (not the expression) with fall-through via goto case,
    // multiple labels on one section, and a default. `goto case` and `goto
    // default` have no Java analogue.
    public string SwitchStatement(int code)
    {
        string label;

        switch (code)
        {
            case 0:
                label = "zero";
                break;

            case 1:
            case 2:
                label = "small";
                break;

            case 3:
                label = "three";
                goto case 4;

            case 4:
                label = "three-or-four";
                break;

            case 5:
                goto default;

            default:
                label = "other";
                break;
        }

        return label;
    }

    // Pattern-matching switch statement with a `when` guard and a var pattern.
    public string SwitchOnPattern(object value)
    {
        switch (value)
        {
            case null:
                return "null";

            case int n when n > 100:
                return "big int";

            case int n:
                return "int " + n;

            case string { Length: 0 }:
                return "empty string";

            case string s:
                return s;

            case var other:
                return other.GetType().Name;
        }
    }

    // Labels, goto, break and continue. C# `goto` is unrestricted; Java has
    // labelled break/continue only.
    public int LabelsAndGoto(int[][] grid)
    {
        int found = -1;

    restart:
        for (int row = 0; row < grid.Length; row++)
        {
            for (int column = 0; column < grid[row].Length; column++)
            {
                if (grid[row][column] < 0)
                {
                    continue;
                }

                if (grid[row][column] == 99)
                {
                    found = row;
                    goto done;
                }

                if (grid[row][column] == 0)
                {
                    break;
                }
            }
        }

        if (found == -2)
        {
            goto restart;
        }

    done:
        return found;
    }

    // Four levels of nesting, each level a different block kind — the port of
    // NestedBlockLinking.java.
    public int DeeplyNested(IEnumerable<int> source)
    {
        int total = 0;

        foreach (int outer in source)
        {
            if (outer % 2 == 0)
            {
                for (int i = 0; i < outer; i++)
                {
                    while (total < 1000)
                    {
                        switch (i % 3)
                        {
                            case 0:
                                total += 1;
                                break;

                            default:
                                {
                                    // A bare block, which is its own scope.
                                    int local = i * outer;
                                    total += local;
                                }

                                break;
                        }
                    }
                }
            }
        }

        return total;
    }

    // A bare block as a statement, introducing a scope with no control flow.
    public int BareBlocks()
    {
        int result = 0;

        {
            int scoped = 1;
            result += scoped;
        }

        {
            // Same name, different scope — legal because the scopes are siblings.
            int scoped = 2;
            result += scoped;
        }

        return result;
    }

    // `lock` — no Java statement form; Java spells it `synchronized`.
    //
    // The lock SUBJECT is its own expression root context, and it was the
    // thinnest in the whole corpus at one row. Every shape a subject can take
    // is below: a field, `this`, a static field, a property, an element, a
    // call result, a local and a chain.
    private static readonly object SharedGate = new object();

    private object GateProperty => gate;

    private readonly object[] gates = { new object(), new object() };

    public int LockBlock(int delta)
    {
        int result;

        lock (gate)
        {
            result = delta;
        }

        lock (this)
        {
            result += 1;
        }

        lock (SharedGate)
        {
            result += 2;
        }

        lock (GateProperty)
        {
            result += 3;
        }

        lock (gates[0])
        {
            result += 4;
        }

        lock (GateFor(delta))
        {
            result += 5;
        }

        object local = gate;
        lock (local)
        {
            result += 6;
        }

        lock (Nested?.gate ?? SharedGate)
        {
            result += 7;
        }

        // Nested locks, which is the shape that deadlocks.
        lock (gate)
        {
            lock (SharedGate)
            {
                result += 8;
            }
        }

        return result;
    }

    private object GateFor(int index) => gates[index % gates.Length];

    public ControlFlowExamples Nested;

    // `checked` / `unchecked` as statements and as expressions.
    public int CheckedAndUnchecked(int large)
    {
        int wrapped;

        unchecked
        {
            wrapped = large * large;
        }

        try
        {
            checked
            {
                wrapped = large + int.MaxValue;
            }
        }
        catch (OverflowException)
        {
            wrapped = 0;
        }

        int expressionForm = unchecked(int.MaxValue + 1);
        return wrapped + expressionForm;
    }

    // `using` statement, `using` with multiple resources, and the C# 8
    // using DECLARATION whose scope is the rest of the enclosing block.
    public string UsingForms(string path)
    {
        using (var reader = new StringReader(path))
        {
            return reader.ReadToEnd();
        }
    }

    public string UsingMultiple(string path)
    {
        using (var first = new StringReader(path))
        using (var second = new StringReader(path))
        {
            return first.ReadToEnd() + second.ReadToEnd();
        }
    }

    public string UsingDeclaration(string path)
    {
        using var reader = new StringReader(path);
        string content = reader.ReadToEnd();
        return content;
    }

    // A block owned by a lambda, and one owned by a local function — both are
    // block scopes whose parent is not a method declaration.
    public int NestedFunctionBlocks(int seed)
    {
        Func<int, int> doubler = x =>
        {
            int inner = x * 2;
            return inner;
        };

        int Tripler(int x)
        {
            int inner = x * 3;
            return inner;
        }

        return doubler(seed) + Tripler(seed);
    }

    public void MonitorBlock()
    {
        bool taken = false;
        try
        {
            Monitor.Enter(gate, ref taken);
        }
        finally
        {
            if (taken)
            {
                Monitor.Exit(gate);
            }
        }
    }
}
